### PR TITLE
Solution cleanup, refactorings and TeamsSkillBot placeholder

### PR DESCRIPTION
**Pull request by Virtual-Josh**
Key changes in this PR

## Deleted old project:
- SsoTeamsHostBot
- TeamsAuthSkillBot
- TeamsSsoSkillBot

## Added TeamsSkillBot placeholder. 
This is just a placeholder for team functionality and should not be ported to other platforms until finished.

## WaterfallHostBot changes

- **RootBot**: Removed unused properties.
- **MainDialog**: Modified dialog so it will prompt for Group of skills (Echo, Waterfall or Teams) and so it will render different skills for each platform after the group selection. It also handles initialization for the special case of SSO dialog.
- **SsoDialog**. Updated it to take connectionName as a parameter. 
- **SsoSignInDialog**: Updated code to display connection name in the prompt.
- Refactored **SkillDefinition** and derived classes in the Skills folder to use enums, make the code lest repetitive and easier to maintain. 
- Added new Group property in **SkillDefintion** (that is also in **appsettings**) to be make it easier to render groups of skills in MainDialog right from config.
- **SkillsConfiguration**: updated it to use Group property in SkillDefintion. 
- **TokenExchangeHandler**: updated internal logic and added additional logging.
- **appsettings.json**: added SsoConnectionNameTeams and removed connectionName (not used), we also added a Group property here and placeholder config entries for TeamsSkillBot

## WaterfallSkillBot changes
- **ActivityRouterDialog**: removed validation for botId (which was wrong) and is not needed if we test without appId and pwd.
- **ActivityRouterDialog**: added new code to handle Update and Delete actions.
- **Cards**: some cleanup and refactoring, however, not this code needs to be reviewed and cleaned up further. 
- **DeleteDialog**: new functionality for testing deletes on the supported channels.
- **FileUploadDialog**: cleanup and upated it to use ConfirmPrompt instead of ChoicePrompt for simple Yes/No prompts.
- **MessageWithAttachmentDialog**: cleanup up, simplified it and removed need to use a controller to simply return an image.
- **UpdateDialog**: new functionality to test simple updates on channels that support it.
- Misc cleanup (removed redundant usings, formatting, use of var, missing (C) headers, etc.)

## Tests updates
- **SubmitAction.json**: fixed typo to make the test pass
- **TaskModule.json**: fixed name to make the test pass
- **WaterfallGreeting.json**: updated it to handle new wording in maindialog test.
- **MessageWithAttachment.json**: updated it to handle new wording. Also updated it to use endsWith() expression to be able to assert part of the returned URL (this was kind of cool and we should consider using more complex expressions in other situations)
- **ProactiveStart.json**: updated it to handle new wording 
- **SignInTests.cs**: fixed bug so it uses the script delivery mode and the right targetskill.
- **SignIn1.json**: updated it to handle new wording 

## Misc
- Added Bots/WaterfallHostBotDotNet.bot file to make it easier to test WaterfallHostBotDotNet from emulator.

### CODE REVIEW

**Review by gabog — CHANGES_REQUESTED** [review]
Hi @Virtual-Josh, I pulled the latest from this branch and the sln is still trying to load some of the deleted projects, can you remove them from the sln file?

image

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Dialogs/ActivityRouterDialog.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            AddDialog(new DeleteDialog());
+            AddDialog(new UpdateDialog());
+
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { ProcessActivityAsync }));
+
+            // The initial child Dialog to run.
+            InitialDialogId = nameof(WaterfallDialog);
+        }
+
+        private async Task<DialogTurnResult> ProcessActivityAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Dialogs/ActivityRouterDialog.cs:51]: There seems to be some overlap between the things the TeamsWaterfall can do and what the WaterfallSkill do, can we remove some of the redundant cases and keep only the things that are special for teams in this one?

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Properties/launchSettings.json [hunk 2]
```diff
@@ -1,4 +1,5 @@
 {
+  "$schema": "http://json.schemastore.org/launchsettings.json",
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Properties/launchSettings.json:2]: Can't make a comment on the actual linke here but the applicationUrl should be set to port 35030

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+﻿# DialogSkillBot
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/README.md:1]: This doesn't seem accurate, remove

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/SimpleGraphClient.cs [hunk 4]
```diff
@@ -8,7 +8,7 @@
 using System.Threading.Tasks;
 using Microsoft.Graph;
 
-namespace Microsoft.BotBuilderSamples
+namespace Microsoft.BotFrameworkFunctionalTests.TeamsWaterfallSkillBot
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/SimpleGraphClient.cs:11]: Where is the simplegraphclient used? I think we should remove it.

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Startup.cs [hunk 5]
```diff
@@ -77,14 +81,17 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+                //.UseWebSockets()
                 .UseRouting()
                 .UseAuthorization()
                 .UseEndpoints(endpoints =>
                 {
                     endpoints.MapControllers();
                 });
 
-            // app.UseHttpsRedirection();
+            // Uncomment this to support HTTPS.
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Startup.cs:93]: This comment seems out of place or refering to a line that is not there, consider removing it.

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj [hunk 6]
```diff
@@ -0,0 +1,47 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <LangVersion>latest</LangVersion>
+    <AssemblyName>Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot</AssemblyName>
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj:6]: The assemble and namespace don't match the project name, Microsoft.BotFrameworkFunctionalTests.TeamsWaterfallSkillBot

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    </Content>
+    <Content Include="Dialogs\Cards\Files\teams-logo.png">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="AdaptiveCards" Version="2.3.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.2" />
+    <PackageReference Include="Microsoft.Bot.Builder.AI.Luis" Version="4.11.1" />
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj:27]: Do we need LUIS in this bot?

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.11.1" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.11.1" />
+    <PackageReference Include="Microsoft.Graph" Version="3.21.0" />
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj:30]: Do we need Graph in this bot?

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    <PackageReference Include="Microsoft.Recognizers.Text.DataTypes.TimexExpression" Version="1.3.2" />
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj:31]: Do we need TimeExpressions in this bot?

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/wwwroot/manifests/waterfallskillbot-manifest-1.0.json [hunk 10]
```diff
@@ -0,0 +1,107 @@
+{
+  "$schema": "https://schemas.botframework.com/schemas/skills/skill-manifest-2.0.0.json",
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/wwwroot/manifests/waterfallskillbot-manifest-1.0.json:2]: This manifest looks old, updated using the one we have in WaterfallSkillBot as a reference.

File: Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Bots/SkillBot.cs [hunk 11]
```diff
@@ -44,6 +44,11 @@ public override async Task OnTurnAsync(ITurnContext turnContext, CancellationTok
             await _conversationState.SaveChangesAsync(turnContext, false, cancellationToken);
         }
 
+        protected override Task<InvokeResponse> OnInvokeActivityAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
```
Comment (gabog) [inline Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Bots/SkillBot.cs:47]: What does this do?

**Review by gabog — CHANGES_REQUESTED** [review]
First pass, I will take another pass once these issues are addressed and I can run the bot.
You should also make sure the functional tests still pass after these changes before commiting.

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Bots/RootBot.cs [hunk 12]
```diff
@@ -15,13 +16,21 @@ namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Bots
     public class RootBot<T> : ActivityHandler
         where T : Dialog
     {
+        public const string ActiveSkillPropertyName = "activeSkillProperty";
+
+        private readonly IStatePropertyAccessor<BotFrameworkSkill> _activeSkillProperty;
         private readonly ConversationState _conversationState;
+        private readonly UserState _userState;
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Bots/RootBot.cs]: removed

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/MainDialog.cs [hunk 13]
```diff
@@ -211,7 +221,7 @@ private async Task<DialogTurnResult> SelectSkillActionStepAsync(WaterfallStepCon
             // Get the skill info based on the selected skill.
             var selectedSkillId = ((FoundChoice)stepContext.Result).Value;
             var deliveryMode = stepContext.Values[_deliveryMode].ToString();
-            var v3Bots = new List<string> { "EchoSkillBotDotNetV3", "EchoSkillBotJSV3" };
+            var v3Bots = new List<string> { "EchoSkillBotV3DotNet", "EchoSkillBotV3JS" };
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/MainDialog.cs]: reverted

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/MainDialog.cs [hunk 14]
```diff
@@ -76,11 +76,21 @@ public MainDialog(ConversationState conversationState, SkillConversationIdFactor
 
             // Add dialog to prepare SSO on the host and test the SSO skill
             // The waterfall skillDialog created in AddSkillDialogs contains the SSO skill action.
-            var waterfallDialog = Dialogs
-                .GetDialogs()
-                .Where(e => e.Id.StartsWith("WaterfallSkill"))
-                .First();
-            AddDialog(new SsoDialog(waterfallDialog, configuration));
+            var waterfallDialog = Dialogs.Find(Constants.WaterfallSkillBotDotNet);
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/MainDialog.cs]: resolved

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj (hunk 8 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj]: we did, but not anymore.

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/wwwroot/manifests/waterfallskillbot-manifest-1.0.json (hunk 10 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/wwwroot/manifests/waterfallskillbot-manifest-1.0.json]: we did, but not anymore.

File: Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Bots/SkillBot.cs (hunk 11 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Bots/SkillBot.cs]: I was testing something and forgot to remove this.

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj (hunk 9 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj]: not anymore

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj (hunk 7 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj]: not anymore

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Bots/RootBot.cs [hunk 15]
```diff
@@ -58,6 +67,23 @@ protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersA
             }
         }
 
+        protected override async Task OnTokenResponseEventAsync(ITurnContext<IEventActivity> turnContext, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Bots/RootBot.cs]: removing.

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Constants.cs [hunk 16]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Constants.cs]: refactored and removed.

File: Bots/DotNet/FunctionalTestsBots.sln [hunk 17]
```diff
@@ -17,8 +17,6 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "CodeFirst", "CodeFirst", "{
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EchoSkillBot-2.1", "Skills\CodeFirst\EchoSkillBot-2.1\EchoSkillBot-2.1.csproj", "{56864219-785B-4600-9849-43CAF90F37E9}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "TeamsSkillBot", "Skills\CodeFirst\TeamsSkillBot\TeamsSkillBot.csproj", "{66E3D426-7627-4D6C-9FAA-87618724F326}"
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/FunctionalTestsBots.sln:20]: updated

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/DeploymentTemplates/LinuxDotNet/template.json [hunk 18]
```diff
@@ -0,0 +1,225 @@
+{
```
[the remark this answers is not in the document]
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/DeploymentTemplates/LinuxDotNet/template.json:1]: removed

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Dialogs/ActivityRouterDialog.cs (hunk 1 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/Dialogs/ActivityRouterDialog.cs:51]: Yes and no. I'll work on building the demo so we can better address this issue.

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/README.md (hunk 3 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/README.md:1]: removed

File: Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj (hunk 6 above)
Comment (Virtual-Josh) [reply Bots/DotNet/Skills/CodeFirst/TeamsWaterfallSkillBot/TeamsWaterfallSkillBot.csproj:6]: updated

**Comment by gabog** [comment]
@ceciliaavila / @denscollo, I apologize for the monster PR, I detailed the main changes in the description. 
Some addtional notes: 
- Don't worry about TeamsSkillBot for now and don't try to port it, it is really not ready.
- Some changes to the wording and the sequence in WaterfallHostBot and WaterfallSkillBot will probably impact the current port to JS, let me know if you have any questions but I think the tests scripts will give you a good hint of where the differences are in addition to the PR description.  

I will wait for your review and approval before merging this PR. 

Thanks

File: Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Dialogs/Update/UpdateDialog.cs [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+            }
+            else
+            {
+                await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Delete is not supported in the {channel} channel."), cancellationToken);
+                return new DialogTurnResult(DialogTurnStatus.Complete);
+            }
+
+            // Ask if we want to update the activity again.
+            var messageText = "Do you want to update the activity again??";
+            var repromptMessageText = "Please select a valid answer";
```
[the remark this answers is not in the document]
Comment (gabog) [reply Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Dialogs/Update/UpdateDialog.cs]: updated it and fixed ?? typo

**Review by denscollo — APPROVED** [review]

**Review by gabog — APPROVED** [review]

**Pull request merged**